crates/phactory/src/system/side_tasks/geo_probe.rs

```diff
@@ -0,0 +1,181 @@
+use chain::BlockNumber;
+use phala_mq::Sr25519MessageChannel;
+
+use crate::side_task::async_side_task::AsyncSideTask;
+use crate::side_task::SideTaskManager;
+
+use std::str::FromStr;
+use std::{error, fmt};
+
+use phala_types::contract;
+use phala_types::messaging::{Geocoding, GeolocationCommand};
+
+use maxminddb::geoip2;
+use std::net::IpAddr;
+
+use crate::secret_channel::SecretMessageChannel;
+use phala_crypto::sr25519::KDF;
+use sp_core::{hashing::blake2_256, sr25519, Pair};
+use std::convert::TryInto;
+
+const BLOCK_INTERVAL: BlockNumber = 2400;
+// 8 hours
+const PROBE_DURATION: BlockNumber = 5; // 1 min
+
+// For detecting the public IP address of worker, we now use the service provided by ipify.org
+// Why we use this service:
+// 1. It is one of the largest and most popular IP address API services on the internet. ipify serves over 30 billion requests per month!
+// 2. It is accessible in both Chinese mainland and other regions.
+// 3. It is open sourced at https://github.com/rdegges/ipify-api.
+const IP_PROBE_URL: &str = "https://api.ipify.org"; // ipinfo is reported to be baned in China
+
+#[derive(Debug)]
+pub enum GeoProbeError {
+    // geo_probe
+    FailedToGetPublicIPAddress,
+    DBNotFound,
+    DBNotValid,
+    IPNotValid,
+    NoRecord,
+    UnknownError,
+}
+
+impl fmt::Display for GeoProbeError {
```

Comment:
@kvinwang I guess there are easier ways to write Errors?

Reply:
> I guess there are easier ways to write Errors?

There is a crate `derive_more` to #[derive(Display)] for no_std and a `thiserror` for std.

Reply:
We already have anyhow as pruntime is like a binary than a library. What do you shuuest?